# Intro to Next Generation Sequencing preprocessing

For these exercises we will be using an dataset sequenced with Illumina MiSeq technology. The organism in question is an enterohaemorrhagic E. coli (EHEC) of the serotype O157, a potentially fatal gastrointestinal pathogen. The sequenced bacterium was part of an outbreak investigation in the St. Louis area, USA in 2011. The sequencing was done as paired-end 2x150bp.

## Set Up
Let's create the directories to organize our data.
```{}
# This will be today's working directory
mkdir ngsintro_exercises && cd ngsintro_exercises
# Directories to organize the data
mkdir Fastq QC Trimmed Mapped Reference
```
## Part 1 - Quality control

### 1. Downloading sequencing data
There are several ways of [downloading](https://ena-docs.readthedocs.io/en/latest/retrieval/file-download.html#using-ena-file-downloader-command-line-tool) sequencing data from different databases. 

The Ecoli raw data that we will use was deposited at the European Nucleotide Archive, under the accession number SRR957824. You could go to the ENA [website](https://www.ebi.ac.uk/ena/browser/) and search for the run with the accession [SRR957824](https://www.ebi.ac.uk/ena/browser/view/SRR957824?show=reads) and download the dataset directly or with the url link for example.

```{}
## Do not run!
wget ftp://ftp.sra.ebi.ac.uk/vol1/fastq/SRR957/SRR957824/SRR957824_2.fastq.gz
```

But this are large datasets, containing around 3 million reads. They will take great time to download and to work with. So instead we will be using a subset of the data:

```{}
```

### 2. Exploring the fastq files
We want to avoid modifying the files by mistake so lets make them unwritable
```{}
cd Fastq
chmod u-w *.fastq*
```
Let's count how many reads we have per file
```{}
zmore *_R1.fastq | grep -E "^@SR" | wc -l 
zmore *_R2.fastq | grep -E "^@SR" | wc -l 
```

### 3. Assessing Quality Control with FastQC
We will use FastQC to assess the quality of our reads. FastQC will produce an html report for each of our files.

Run fastQC in each of our fastq files
```{}
# To go back to our main wd
cd ..

fastqc --outdir QC/ Fastq/SRR957824_500K_R1.fastq.gz
fastqc --outdir QC/ Fastq/SRR957824_500K_R2.fastq.gz
```

Alternatively, you can also tell fastqc to take all of the fastq files as input
```{}
fastqc --outdir QC/ Fastq/*.fastq*
```

Let's inspect the QC reports generated by fastQC.

What do you see? Is there any property that could cause bias in the downstream analyses?

### 4. Trim and filter with CutAdapt
After inspecting and interpreting our quality plots, we know that there is still some adapters in our reads. 

We will remove the adapters and trim low-quality bases from 5' ends of each
read before adapter removal. 

```{}
GATCGGAAGAGCACACGTCTGAACTCCAGTCACAGCGCTATCTCGTATGC

cutadapt -q 25 -a AGATCGGAAGAGCACACGTCTGAACTCCAGTCA -A AGATCGGAAGAGCGTCGTGTAGGGAAAGAGTGT --minimum-length 50 -o Trimmed/SRR957824_R1.trimmed.fastq.gz -p SRR957824_R1.trimmed.fastq.gz Fastq/SRR957824_500K_R1.fastq.gz Fastq/SRR957824_500K_R2.fastq.gz
```

Lets run fastQC again

```{}
fastqc --outdir ../qc/ *.fastq*
```
### 5. Combining Reports in one with MultiQC
Imagine you have a lot of fastq files. Going through each of the fastqc reports individually before AND after filtering is time consuming and difficult to track.

MultiQC is a tool that can recopilate reports from different bioinformatics tools and from multiple files into a single report.

```{}
multiqc *
```

## Part 2 - Mapping to a Reference Genome

### 6. Getting the  Reference Genome
We are going to map our reads to the E. coli O157:H7 str. Sakai Genome with accession number [ASM886v2](https://www.ncbi.nlm.nih.gov/assembly/GCF_000008865.2) in NCBI.

```{}
wget https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/000/008/865/GCF_000008865.2_ASM886v2/GCF_000008865.2_ASM886v2_genomic.fna.gz
```

Before aligning the reads against a reference, it is necessary to build an index of that reference. 

Indexing the reference is a necessary pre-processing step that makes searching for patterns much much faster. Many popular aligners such as Bowtie and BWA use an algorithm called the Burrowsâ€“Wheeler transform to build the index.

Let's Index the reference genome.
```{}
bowtie2-build GCF_000008865.2_ASM886v2_genomic.fna.gz GCF_000008865
```
### 7. Mapping reads to the Reference using Bowtie2

```{}
bowtie2 -x Reference/GCA_000471505 -1 Trimmed/R1_clean.fastq.gz  -2 Trimmed/R2_clean.fastq.gz -S Mapped/SRR957824.sam
```
<details>
    <summary>Bowtie2 Output</summary>
        
        419066 reads; of these:
          419066 (100.00%) were paired; of these:
          47026 (11.22%) aligned concordantly 0 times
          343686 (82.01%) aligned concordantly exactly 1 time
          28354 (6.77%) aligned concordantly >1 times
        ----
        47026 pairs aligned concordantly 0 times; of these:
          36851 (78.36%) aligned discordantly 1 time
        ----
        10175 pairs aligned 0 times concordantly or discordantly; of these:
            20350 mates make up the pairs; of these:
            12649 (62.16%) aligned 0 times
            1182 (5.81%) aligned exactly 1 time
            6519 (32.03%) aligned >1 times
    98.49% overall alignment rate
</details>

Let's inspect our Sam file output!
```{}
head sam
tail sam
```

### 7. Using Samtools: Sam to Bam, sorting and indexing
Doing a head/tail is not very informative, but we can use samtools to manage our sam/bam files.
```{}
samtools view -hbo SRR957824.bam SRR957824.sam
```
We can now sort by position and index our bam file.
```{}
samtools sort SRR957824.bam > SRR957824.sorted.bam
samtools index SRR957824.sorted.bam
```

### 8. Mapping Stats
```{}
samtools stats SRR957824.sorted.bam > SRR957824.stats
```
```{}
multiqc SRR957824.stats
```

### 9. Removing Duplicates
```{}
samtools fixmate <in.nameSrt.bam> <out.bam>
samtools rmdup <input.bam> SRR957824.rmdups.bam
samtools index SRR957824.rmdups.bam
```

### 10. Alignment metrics
Since the BAM file contains records for both mapped and unmapped reads, just counting records doesn't provide information about the mapping rate of our alignment. The samtools flagstat tool provides a simple analysis of mapping rate based on the the SAM flag fields.
```{}
samtools flagstat SRR957824.rmdups.bam
```

Samtools idxstats report shows how many reads aligned to each contig in your reference.
```{}
samtools idxstats SRR957824.rmdups.bam
```

The output has four tab-delimited columns:
1. contig name
2. contig length
3. number of mapped reads
4. number of unmapped reads

If you're mapping to a non-genomic reference such as miRBase miRNAs or another set of genes (a transcriptome), samtools idxstats gives you a quick look at quantitative alignment results.




