<!DOCTYPE html>
<html>
<head>
<title>NGS_intro_handson.md</title>
<meta http-equiv="Content-type" content="text/html;charset=UTF-8">

<style>
/* https://github.com/microsoft/vscode/blob/master/extensions/markdown-language-features/media/markdown.css */
/*---------------------------------------------------------------------------------------------
 *  Copyright (c) Microsoft Corporation. All rights reserved.
 *  Licensed under the MIT License. See License.txt in the project root for license information.
 *--------------------------------------------------------------------------------------------*/

body {
	font-family: var(--vscode-markdown-font-family, -apple-system, BlinkMacSystemFont, "Segoe WPC", "Segoe UI", "Ubuntu", "Droid Sans", sans-serif);
	font-size: var(--vscode-markdown-font-size, 14px);
	padding: 0 26px;
	line-height: var(--vscode-markdown-line-height, 22px);
	word-wrap: break-word;
}

#code-csp-warning {
	position: fixed;
	top: 0;
	right: 0;
	color: white;
	margin: 16px;
	text-align: center;
	font-size: 12px;
	font-family: sans-serif;
	background-color:#444444;
	cursor: pointer;
	padding: 6px;
	box-shadow: 1px 1px 1px rgba(0,0,0,.25);
}

#code-csp-warning:hover {
	text-decoration: none;
	background-color:#007acc;
	box-shadow: 2px 2px 2px rgba(0,0,0,.25);
}

body.scrollBeyondLastLine {
	margin-bottom: calc(100vh - 22px);
}

body.showEditorSelection .code-line {
	position: relative;
}

body.showEditorSelection .code-active-line:before,
body.showEditorSelection .code-line:hover:before {
	content: "";
	display: block;
	position: absolute;
	top: 0;
	left: -12px;
	height: 100%;
}

body.showEditorSelection li.code-active-line:before,
body.showEditorSelection li.code-line:hover:before {
	left: -30px;
}

.vscode-light.showEditorSelection .code-active-line:before {
	border-left: 3px solid rgba(0, 0, 0, 0.15);
}

.vscode-light.showEditorSelection .code-line:hover:before {
	border-left: 3px solid rgba(0, 0, 0, 0.40);
}

.vscode-light.showEditorSelection .code-line .code-line:hover:before {
	border-left: none;
}

.vscode-dark.showEditorSelection .code-active-line:before {
	border-left: 3px solid rgba(255, 255, 255, 0.4);
}

.vscode-dark.showEditorSelection .code-line:hover:before {
	border-left: 3px solid rgba(255, 255, 255, 0.60);
}

.vscode-dark.showEditorSelection .code-line .code-line:hover:before {
	border-left: none;
}

.vscode-high-contrast.showEditorSelection .code-active-line:before {
	border-left: 3px solid rgba(255, 160, 0, 0.7);
}

.vscode-high-contrast.showEditorSelection .code-line:hover:before {
	border-left: 3px solid rgba(255, 160, 0, 1);
}

.vscode-high-contrast.showEditorSelection .code-line .code-line:hover:before {
	border-left: none;
}

img {
	max-width: 100%;
	max-height: 100%;
}

a {
	text-decoration: none;
}

a:hover {
	text-decoration: underline;
}

a:focus,
input:focus,
select:focus,
textarea:focus {
	outline: 1px solid -webkit-focus-ring-color;
	outline-offset: -1px;
}

hr {
	border: 0;
	height: 2px;
	border-bottom: 2px solid;
}

h1 {
	padding-bottom: 0.3em;
	line-height: 1.2;
	border-bottom-width: 1px;
	border-bottom-style: solid;
}

h1, h2, h3 {
	font-weight: normal;
}

table {
	border-collapse: collapse;
}

table > thead > tr > th {
	text-align: left;
	border-bottom: 1px solid;
}

table > thead > tr > th,
table > thead > tr > td,
table > tbody > tr > th,
table > tbody > tr > td {
	padding: 5px 10px;
}

table > tbody > tr + tr > td {
	border-top: 1px solid;
}

blockquote {
	margin: 0 7px 0 5px;
	padding: 0 16px 0 10px;
	border-left-width: 5px;
	border-left-style: solid;
}

code {
	font-family: Menlo, Monaco, Consolas, "Droid Sans Mono", "Courier New", monospace, "Droid Sans Fallback";
	font-size: 1em;
	line-height: 1.357em;
}

body.wordWrap pre {
	white-space: pre-wrap;
}

pre:not(.hljs),
pre.hljs code > div {
	padding: 16px;
	border-radius: 3px;
	overflow: auto;
}

pre code {
	color: var(--vscode-editor-foreground);
	tab-size: 4;
}

/** Theming */

.vscode-light pre {
	background-color: rgba(220, 220, 220, 0.4);
}

.vscode-dark pre {
	background-color: rgba(10, 10, 10, 0.4);
}

.vscode-high-contrast pre {
	background-color: rgb(0, 0, 0);
}

.vscode-high-contrast h1 {
	border-color: rgb(0, 0, 0);
}

.vscode-light table > thead > tr > th {
	border-color: rgba(0, 0, 0, 0.69);
}

.vscode-dark table > thead > tr > th {
	border-color: rgba(255, 255, 255, 0.69);
}

.vscode-light h1,
.vscode-light hr,
.vscode-light table > tbody > tr + tr > td {
	border-color: rgba(0, 0, 0, 0.18);
}

.vscode-dark h1,
.vscode-dark hr,
.vscode-dark table > tbody > tr + tr > td {
	border-color: rgba(255, 255, 255, 0.18);
}

</style>

<style>
/* Tomorrow Theme */
/* http://jmblog.github.com/color-themes-for-google-code-highlightjs */
/* Original theme - https://github.com/chriskempson/tomorrow-theme */

/* Tomorrow Comment */
.hljs-comment,
.hljs-quote {
	color: #8e908c;
}

/* Tomorrow Red */
.hljs-variable,
.hljs-template-variable,
.hljs-tag,
.hljs-name,
.hljs-selector-id,
.hljs-selector-class,
.hljs-regexp,
.hljs-deletion {
	color: #c82829;
}

/* Tomorrow Orange */
.hljs-number,
.hljs-built_in,
.hljs-builtin-name,
.hljs-literal,
.hljs-type,
.hljs-params,
.hljs-meta,
.hljs-link {
	color: #f5871f;
}

/* Tomorrow Yellow */
.hljs-attribute {
	color: #eab700;
}

/* Tomorrow Green */
.hljs-string,
.hljs-symbol,
.hljs-bullet,
.hljs-addition {
	color: #718c00;
}

/* Tomorrow Blue */
.hljs-title,
.hljs-section {
	color: #4271ae;
}

/* Tomorrow Purple */
.hljs-keyword,
.hljs-selector-tag {
	color: #8959a8;
}

.hljs {
	display: block;
	overflow-x: auto;
	color: #4d4d4c;
	padding: 0.5em;
}

.hljs-emphasis {
	font-style: italic;
}

.hljs-strong {
	font-weight: bold;
}
</style>

<style>
/*
 * Markdown PDF CSS
 */

 body {
	font-family: -apple-system, BlinkMacSystemFont, "Segoe WPC", "Segoe UI", "Ubuntu", "Droid Sans", sans-serif, "Meiryo";
	padding: 0 12px;
}

pre {
	background-color: #f8f8f8;
	border: 1px solid #cccccc;
	border-radius: 3px;
	overflow-x: auto;
	white-space: pre-wrap;
	overflow-wrap: break-word;
}

pre:not(.hljs) {
	padding: 23px;
	line-height: 19px;
}

blockquote {
	background: rgba(127, 127, 127, 0.1);
	border-color: rgba(0, 122, 204, 0.5);
}

.emoji {
	height: 1.4em;
}

code {
	font-size: 14px;
	line-height: 19px;
}

/* for inline code */
:not(pre):not(.hljs) > code {
	color: #C9AE75; /* Change the old color so it seems less like an error */
	font-size: inherit;
}

/* Page Break : use <div class="page"/> to insert page break
-------------------------------------------------------- */
.page {
	page-break-after: always;
}

</style>

<script src="https://unpkg.com/mermaid/dist/mermaid.min.js"></script>
</head>
<body>
  <script>
    mermaid.initialize({
      startOnLoad: true,
      theme: document.body.classList.contains('vscode-dark') || document.body.classList.contains('vscode-high-contrast')
          ? 'dark'
          : 'default'
    });
  </script>
<h1 id="intro-to-next-generation-sequencing-preprocessing">Intro to Next Generation Sequencing preprocessing</h1>
<h3 id="mcgill-initiative-in-computational-medicine-summer-2022">McGill Initiative in Computational Medicine, Summer 2022</h3>
<hr>
<p>For these exercises we will be using an dataset sequenced with Illumina MiSeq technology. The organism in question is an enterohaemorrhagic E. coli (EHEC) of the serotype O157, a potentially fatal gastrointestinal pathogen. The sequenced bacterium was part of an outbreak investigation in the St. Louis area, USA in 2011.</p>
<h2 id="set-up">Set Up</h2>
<p>Let's create the directories to organize our data.</p>
<pre class="hljs"><code><div>## This will be today's working directory
mkdir ngsintro_exercises &amp;&amp; cd ngsintro_exercises

## Create directories to organize the data
mkdir Fastq QC Trimmed Mapped Reference
</div></code></pre>
<h2 id="part-1---quality-control">Part 1 - Quality control</h2>
<h3 id="1-downloading-sequencing-data">1. Downloading sequencing data</h3>
<p>There are several ways of <a href="https://ena-docs.readthedocs.io/en/latest/retrieval/file-download.html#using-ena-file-downloader-command-line-tool">downloading</a> sequencing data from different databases.</p>
<p>The Ecoli raw data that we will use was deposited at the European Nucleotide Archive, under the accession number SRR957824. You could go to the ENA <a href="https://www.ebi.ac.uk/ena/browser/">website</a> and search for the run with the accession <a href="https://www.ebi.ac.uk/ena/browser/view/SRR957824?show=reads">SRR957824</a> and download the dataset directly or with the url link for example.</p>
<pre class="hljs"><code><div>## Do not run!
wget ftp://ftp.sra.ebi.ac.uk/vol1/fastq/SRR957/SRR957824/SRR957824_2.fastq.gz
</div></code></pre>
<p>But this are large datasets, containing around 3 million reads. They will take great time to download and to work with. So instead we will be using a subset of the data:</p>
<p>Download the data from this <a href="https://mcgill-my.sharepoint.com/:f:/g/personal/maria_femerlingromero_mail_mcgill_ca/EuAYhtMeLJdPtwHxPAHrIgMBLGGon-r52KlxgHP2oTo7xA?e=uSCoeV">link</a></p>
<p>Or download directly from the terminal in the Fastq directory.</p>
<pre class="hljs"><code><div>cd Fastq
curl -O -J -L https://osf.io/shqpv/download
curl -O -J -L https://osf.io/9m3ch/download
</div></code></pre>
<h3 id="2-exploring-the-fastq-files">2. Exploring the fastq files</h3>
<p>We want to avoid modifying the files by mistake so lets make them unwritable</p>
<pre class="hljs"><code><div>cd Fastq
chmod u-w *.fastq*
</div></code></pre>
<p>Judging from the file names, what type of reads do we
have?</p>
<p>Let's see a bit of the format</p>
<pre class="hljs"><code><div>zmore *_R1.fastq.gz | head
</div></code></pre>
<p>Do you think this was short-read or long read?</p>
<p>Let's count how many reads we have per file</p>
<pre class="hljs"><code><div>zmore *_R1.fastq.gz | grep -E &quot;^@SR&quot; | wc -l 
zmore *_R2.fastq.gz | grep -E &quot;^@SR&quot; | wc -l 
</div></code></pre>
<p>There are 500K reads on each file.
Does that mean that we have 1 Million independent reads?</p>
<h3 id="3-assessing-quality-control-with-fastqc">3. Assessing Quality Control with FastQC</h3>
<p>We will use FastQC to assess the quality of our reads. FastQC will produce an html report for each of our files.</p>
<p>To explore FastQC options you can run</p>
<pre class="hljs"><code><div>fastqc --help
</div></code></pre>
<p>Run fastQC in each of our fastq files</p>
<pre class="hljs"><code><div># To go back to our main wd
cd ..

fastqc --outdir QC/ Fastq/SRR957824_500K_R1.fastq.gz
fastqc --outdir QC/ Fastq/SRR957824_500K_R2.fastq.gz
</div></code></pre>
<p>Alternatively, you can also tell fastqc to take all of the fastq files as input</p>
<pre class="hljs"><code><div>fastqc --outdir QC/ Fastq/*.fastq*
</div></code></pre>
<p>Let's inspect the QC reports generated by fastQC.</p>
<p>What do you see? Is there any property that could cause bias in the downstream analyses?</p>
<h2 id="part-2---improving-the-quality-of-reads">Part 2 - Improving the Quality of Reads</h2>
<h3 id="4-trim-and-filter-with-cutadapt">4. Trim and filter with CutAdapt</h3>
<p>After inspecting and interpreting our quality plots, we know that there is still some adapters in our reads.</p>
<p>We will use cutadapt to trim and filter our reads.</p>
<p>You can inspect the options by running --help.</p>
<pre class="hljs"><code><div>cutadapt --help
</div></code></pre>
<p>We want to remove the illumina adapters and trim low-quality bases from 5' ends of each paired read before adapter removal. So let's give those options to cutadapt.</p>
<blockquote>
<p>It is a convention to add to the names whatever process/filtering was done. In this case we will add _<em>trimmed</em> to indicate that we have removed adapters and trimmed ends off our fastq files.</p>
</blockquote>
<pre class="hljs"><code><div>cutadapt -q 25 -a AGATCGGAAGAGCACACGTCTGAACTCCAGTCA -A AGATCGGAAGAGCGTCGTGTAGGGAAAGAGTGT --minimum-length 50 -o Trimmed/SRR957824_R1_trimmed.fastq.gz -p Trimmed/SRR957824_R2_trimmed.fastq.gz Fastq/SRR957824_500K_R1.fastq.gz Fastq/SRR957824_500K_R2.fastq.gz &gt; Trimmed/cutadapt.log
</div></code></pre>
<p>Why do you think we need to run both paired-end files together?</p>
<p>Lets run fastQC again</p>
<pre class="hljs"><code><div>fastqc --outdir QC/ Trimmed/*trimmed.fastq*
</div></code></pre>
<h3 id="5-combining-reports-in-one-with-multiqc">5. Combining Reports in one with MultiQC</h3>
<p>Imagine you have a lot of fastq files. Going through each of the fastqc reports individually before AND after filtering is time consuming and difficult to track.</p>
<p>MultiQC is a tool that can recopilate reports from different bioinformatics tools and from multiple files into a single report.</p>
<pre class="hljs"><code><div>multiqc QC --outdir QC
</div></code></pre>
<h2 id="part-3---mapping-to-a-reference-genome">Part 3 - Mapping to a Reference Genome</h2>
<h3 id="6-getting-the-reference-genome">6. Getting the  Reference Genome</h3>
<p>We are going to map our reads to the E. coli O157:H7 str. Sakai Genome with accession number <a href="https://www.ncbi.nlm.nih.gov/assembly/GCF_000008865.2">ASM886v2</a> in NCBI.</p>
<p>You can download the fasta from the previous link or directly using wget.</p>
<pre class="hljs"><code><div>cd Reference
wget https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/000/008/865/GCF_000008865.2_ASM886v2/GCF_000008865.2_ASM886v2_genomic.fna.gz
</div></code></pre>
<p>Before aligning the reads against a reference, it is necessary to build an index of that reference.</p>
<p>Indexing the reference is a necessary pre-processing step that makes searching for patterns much much faster. Many popular aligners such as Bowtie and BWA use an algorithm called the Burrows–Wheeler transform to build the index.</p>
<p>Let's Index the reference genome.</p>
<pre class="hljs"><code><div>bowtie2-build GCF_000008865.2_ASM886v2_genomic.fna.gz GCF_000008865
</div></code></pre>
<h3 id="7-mapping-reads-to-the-reference-using-bowtie2">7. Mapping reads to the Reference using Bowtie2</h3>
<p>Bowtie2 is a splice-unaware aligner used to map sequencing reads to long reference genomes.</p>
<pre class="hljs"><code><div>cd ..
bowtie2 --help
</div></code></pre>
<pre class="hljs"><code><div>bowtie2 -x Reference/GCF_000008865 -1 Trimmed/SRR957824_R1_trimmed.fastq.gz  -2 Trimmed/SRR957824_R2_trimmed.fastq.gz -S Mapped/SRR957824.sam
</div></code></pre>
<details>
    <summary>Bowtie2 Output</summary>
<pre><code>  414756 reads; of these:
    414756 (100.00%) were paired; of these:
      46689 (11.26%) aligned concordantly 0 times
      340122 (82.01%) aligned concordantly exactly 1 time
      27945 (6.74%) aligned concordantly &gt;1 times
      ----
      46689 pairs aligned concordantly 0 times; of these:
        36612 (78.42%) aligned discordantly 1 time
      ----
      10077 pairs aligned 0 times concordantly or discordantly; of these:
        20154 mates make up the pairs; of these:
          12534 (62.19%) aligned 0 times
          1162 (5.77%) aligned exactly 1 time
          6458 (32.04%) aligned &gt;1 times
  98.49% overall alignment rate
</code></pre>
</details>
<p>Let's inspect our Sam file output!</p>
<pre class="hljs"><code><div>cd Mapped
head SRR957824.sam
</div></code></pre>
<p>Can you identify the CIGAR string of the first mapped read?</p>
<h3 id="7-using-samtools-sam-to-bam-sorting-and-indexing">7. Using Samtools: Sam to Bam, sorting and indexing</h3>
<p>Doing a head/tail is not very informative, but we can use samtools to manage our sam/bam files.</p>
<p>Samtools is a software package for parsing and manipulating alignments in the SAM/BAM format.</p>
<p>The first step is to compress the sam file to a bam file.</p>
<blockquote>
<p>Generally, try not to keep alignments in sam format as they can get very <em>very</em> large and will consume a lot of precious storage space!</p>
</blockquote>
<p>To convert a sam to a bam we use samtools view.</p>
<pre class="hljs"><code><div>samtools view -hbo SRR957824.bam SRR957824.sam
</div></code></pre>
<h3 id="8-marking-duplicates">8. Marking Duplicates</h3>
<p>Duplicates are sets of reads pairs that have the same unclipped alignment start
and unclipped alignment end. They’re suspected to be non-independent measurements of a sequence.</p>
<p>We deal with duplicated reads by marking them using samtools or picard.</p>
<pre class="hljs"><code><div>samtools fixmate -m SRR957824.bam SRR957824.fixmate.bam
</div></code></pre>
<p>We can now sort by position our bam file and mark the duplicated reads. Removing the duplicated reads is optional and can be performed by adding the <em>-r</em> option to samtools markdup.</p>
<pre class="hljs"><code><div>samtools sort -o SRR957824.fixmate.sorted.bam SRR957824.fixmate.bam 
samtools markdup -r SRR957824.fixmate.sorted.bam SRR957824.markdup.bam
</div></code></pre>
<p>To finish, we need to index our bam file.</p>
<p>Indexing a genome sorted BAM file allows one to quickly extract alignments overlapping particular genomic regions and is required for most visualizers like IGV.</p>
<pre class="hljs"><code><div>samtools index SRR957824.markdup.bam
</div></code></pre>
<h3 id="9-alignment-metrics">9. Alignment metrics</h3>
<p>Since the BAM file contains records for both mapped and unmapped reads, just counting records doesn't provide information about the mapping rate of our alignment. The samtools flagstat tool provides a simple analysis of mapping rate based on the the SAM flag fields.</p>
<p>Let's see out alingment stats:</p>
<pre class="hljs"><code><div>samtools flagstat SRR957824.markdup.bam
</div></code></pre>
<blockquote>
<p>Exercise: run samtools flagstat in the file we had before removing the duplicates.</p>
</blockquote>
<details>
    <summary>Flagstat Output</summary>
<pre><code>    - Before Marking and removing duplicates -
    829512 + 0 in total (QC-passed reads + QC-failed reads)
    829512 + 0 primary
    0 + 0 secondary
    0 + 0 supplementary
    0 + 0 duplicates
    0 + 0 primary duplicates
    816978 + 0 mapped (98.49% : N/A)
    816978 + 0 primary mapped (98.49% : N/A)
    829512 + 0 paired in sequencing
    414756 + 0 read1
    414756 + 0 read2
    736134 + 0 properly paired (88.74% : N/A)
    815700 + 0 with itself and mate mapped
    1278 + 0 singletons (0.15% : N/A)
    104 + 0 with mate mapped to a different chr
    13 + 0 with mate mapped to a different chr (mapQ&gt;=5)

    - After Marking and removing duplicates -
    828902 + 0 in total (QC-passed reads + QC-failed reads)
    828902 + 0 primary
    0 + 0 secondary
    0 + 0 supplementary
    0 + 0 duplicates
    0 + 0 primary duplicates
    816368 + 0 mapped (98.49% : N/A)
    816368 + 0 primary mapped (98.49% : N/A)
    828902 + 0 paired in sequencing
    414444 + 0 read1
    414458 + 0 read2
    735672 + 0 properly paired (88.75% : N/A)
    815196 + 0 with itself and mate mapped
    1172 + 0 singletons (0.14% : N/A)
    104 + 0 with mate mapped to a different chr
    13 + 0 with mate mapped to a different chr (mapQ&gt;=5)
</code></pre>
</details>
<p>Can you infer how many duplicates we had in out data by comparing the stats?</p>
<p>Samtools idxstats report shows how many reads aligned to each contig in your reference.</p>
<pre class="hljs"><code><div>samtools idxstats SRR957824.markdup.bam
</div></code></pre>
<details>
    <summary>IdxStats Output</summary>
<pre><code>    NC_002695.2     5498578 798728  1216
    NC_002127.1     3306    0       0
    NC_002128.1     92721   17640   62
            *       0       0       11256
</code></pre>
</details>
<p>The output has four tab-delimited columns:</p>
<ol>
<li>contig name</li>
<li>contig length</li>
<li>number of mapped reads</li>
<li>number of unmapped reads</li>
</ol>
<p>If you're mapping to a non-genomic reference such as miRBase miRNAs or another set of genes (a transcriptome, samtools idxstats gives you a quick look at quantitative alignment results.</p>
<h3 id="10-more-filters-with-samtools">10. More filters with Samtools</h3>
<p>Let's say you are only interested in a certain region, or need only reads with a certain flag. Samtools allows you to filter bam/sam files.</p>
<p>Extract only reads that mapped to the region from 18000bp-18050bp</p>
<pre class="hljs"><code><div>samtools view SRR957824.markdup.bam &quot;NC_002695.2:18000-18050&quot;
</div></code></pre>
<p>Extract only the unmapped reads.</p>
<pre class="hljs"><code><div>samtools view -f 4 SRR957824.markdup.bam | head
</div></code></pre>
<p>Samtools is a very extensive software package!
Just in samtools you can:</p>
<ul>
<li>Filter</li>
<li>Get mapping stats</li>
<li>Call variants</li>
<li>Split bam</li>
<li>and much more!</li>
</ul>
<p>I invite you to see the <a href="http://www.htslib.org/doc/samtools.html">documentation</a> to know more about the tools available.</p>
<hr>
<p>Markdown and exercises prepared by Georgette Femerling.</p>

</body>
</html>
